panther_hardware_interfaces/include/panther_hardware_interfaces/panther_system.hpp

```diff
@@ -0,0 +1,136 @@
+  std::vector<CommandInterface> export_command_interfaces() override;
+
+  return_type read(const rclcpp::Time & time, const rclcpp::Duration & period) override;
+  return_type write(const rclcpp::Time & time, const rclcpp::Duration & period) override;
+
+protected:
+  void CheckJointSize();
+  void SortJointNames();
+  void CheckJointNames();
+  void SetInitialValues();
+  void CheckInterfaces();
+  void ReadDrivetrainSettings();
+  void ReadCanOpenSettings();
+  void ReadInitializationActivationAttempts();
+  void ReadParametersAndCreateRoboteqErrorFilter();
+
+  void UpdateHwStates();
+  void UpdateDriverState();
+  void UpdateSystemFeedback();
+  void UpdateMsgErrors();
+
+  static constexpr size_t kJointsSize = 4;
+
+  // Currently only velocity command mode is supported - although Roboteq driver support position
+  // and torque mode, in 2.1 firmware both modes aren't really stable and safe.
+  // In torque mode sometimes after killing the software motor moves and it generally isn't well
+  // tuned. Position mode also isn't really stable (reacts abruptly to spikes). If updating the
+  // firmware to 2.1a will solve these issues, it may be worth to enable other modes.
+  double hw_commands_velocities_[kJointsSize];
+
+  double hw_states_positions_[kJointsSize];
+  double hw_states_velocities_[kJointsSize];
+  double hw_states_efforts_[kJointsSize];
+
+  // Define expected joint order, so that it doesn't matter what order is defined in the URDF.
+  // It is expected that the joint name should contain these specifiers.
+  std::string joint_order_[kJointsSize] = {"fl", "fr", "rl", "rr"};
+  std::string joints_names_sorted_[kJointsSize];
+
+  std::unique_ptr<GPIOController> gpio_controller_;
+  std::shared_ptr<MotorsController> motors_controller_;
+
+  DrivetrainSettings drivetrain_settings_;
+  CanOpenSettings canopen_settings_;
+
+  PantherSystemRosInterface panther_system_ros_interface_;
+
+  // Sometimes SDO errors can happen during initialization and activation of Roboteq drivers,
+  // in these cases it is better to retry
+  // Example errors:
+  // SDO abort code 05040000 received on upload request of object 1000 (Device type) to
+  // node 02: SDO protocol timed out
+  // SDO abort code 05040000 received on upload request of sub-object 1018:01 (Vendor-ID) to
+  // node 02: SDO protocol timed out
+  unsigned max_roboteq_initialization_attempts_ = 2;
+  unsigned max_roboteq_activation_attempts_ = 2;
+
+  // SDO error can happen also during setting safety stop (it may be not necessary to use attempts
+  // once we have GPIO controller)
+  unsigned max_safety_stop_attempts_ = 20;
+
+  rclcpp::Logger logger_{rclcpp::get_logger("PantherSystem")};
+  rclcpp::Clock steady_clock_{RCL_STEADY_TIME};
+
+  std::shared_ptr<RoboteqErrorFilter> roboteq_error_filter_;
+  struct ErrorsFilterIds
+  {
+    std::size_t read_sdo;
+    std::size_t write_sdo;
+    std::size_t read_pdo;
+    std::size_t roboteq_driver;
+  } errors_filter_ids_;
```

Comment:
```suggestion
  enum ErrorsFilterIds
  {
    READ_SDO = 0,
    WRITE_PDO = 1,
    READ_PDO = 2,
    ROBOTEQ_DRIVER = 3
  };
```
This should be enum, not a structure. Also since the IDs are not changed anywhere they should be treated as constants, hence written all in capital letters. For example replacing
```c++
errors_filter_ids_.read_sdo;
```
with
```c++
ErrorsFilterIds::READ_SDO;
```

Reply:
Yes, definitely it should be enum.